File: src/engine.js
import { fromIso } from "./clock.js";

function getThresholds(zone, crop) {
  const mn = zone.min_pct !== null && zone.min_pct !== undefined ? zone.min_pct : crop.min_pct;
  const tg = zone.target_pct !== null && zone.target_pct !== undefined ? zone.target_pct : crop.target_pct;
  return [mn, tg];
}

function rainExpected(weather, nowIso, cfg) {
  if (!weather || !weather.hourly_times) return [false, 0.0, 0.0];
  let total = 0.0, maxProb = 0.0, count = 0;
  for (let i = 0; i < weather.hourly_times.length; i++) {
    if (weather.hourly_times[i] < nowIso) continue;
    if (count >= cfg.engine.rain_lookahead_hours) break;
    total += weather.hourly_precip_mm[i] || 0.0;
    maxProb = Math.max(maxProb, weather.hourly_precip_prob[i] || 0.0);
    count++;
  }
  const byVol = total >= cfg.engine.rain_skip_mm;
  const byProb = maxProb >= cfg.engine.rain_skip_prob && total >= cfg.engine.rain_skip_prob_min_mm;
  return [byVol || byProb, total, maxProb];
}

function slopeHours(readings, mn, moisture) {
  if (readings.length < 4) return null;
  const pts = readings.slice(0, 12).map(r => r.moisture_pct).reverse();
  const n = pts.length;
  let sx = 0, sy = 0, sxy = 0, sxx = 0;
  for (let i = 0; i < n; i++) {
    sx += i; sy += pts[i]; sxy += i * pts[i]; sxx += i * i;
  }
  const denom = n * sxx - sx * sx;
  if (denom === 0) return null;
  const slope = (n * sxy - sx * sy) / denom;
  if (slope >= 0) return null;
  return (moisture - mn) / Math.abs(slope) * (5.0 / 60.0);
}

function et0Hours(weather, crop, moisture, mn) {
  if (!weather || !weather.hourly_evaporation_mm) return null;
  const vals = weather.hourly_evaporation_mm.filter(v => v !== null && v !== undefined);
  if (vals.length === 0) return null;
  const mean = vals.reduce((a, b) => a + b, 0) / vals.length;
  const rate = mean * crop.crop_factor / crop.water_holding_mm * 100.0;
  return rate > 0 ? (moisture - mn) / rate : null;
}

export function duration(moisture, tg, whMm, flow, eff, maxMin) {
  const mins = (tg - moisture) / 100.0 * whMm / (flow * eff) * 60.0;
  if (mins <= 0) return [0.0, false];
  return mins > maxMin ? [maxMin, true] : [mins, false];
}

function nextWindow(localMin, windows) {
  const parsed = windows.map(w => {
    const s = parseInt(w.slice(0, 2), 10) * 60 + parseInt(w.slice(3, 5), 10);
    const e = parseInt(w.slice(6, 8), 10) * 60 + parseInt(w.slice(9, 11), 10);
    return [s, e];
  });
  for (const [s, e] of parsed) {
    if (s <= localMin && localMin <= e) return [true, null];
  }
  const starts = parsed.map(p => p[0]).sort((a, b) => a - b);
  const nxt = starts.find(s => s > localMin) ?? starts[0];
  const hh = String(Math.floor(nxt / 60)).padStart(2, "0");
  const mm = String(nxt % 60).padStart(2, "0");
  return [false, `${hh}:${mm}`];
}

export function decide(zone, crop, readings, weather, nowIso, localMin, cfg) {
  if (!crop) throw new Error(`Zone ${zone.id}: unknown crop`);
  const [mn, tg] = getThresholds(zone, crop);
  if (mn >= tg) throw new Error(`Zone ${zone.id}: min_pct must be less than target_pct`);
  const crit = mn - cfg.engine.critical_margin_pts;
  const trace = [`min=${mn}% target=${tg}% critical=${crit.toFixed(1)}%`];
  const eff = cfg.efficiency[zone.irrigation];
  if (!readings || readings.length === 0) {
    return { action: "NO_DATA", minutes: null, reason: "no sensor data", hours_until: null, trace };
  }
  const age = (fromIso(nowIso) - fromIso(readings[0].ts)) / 60000.0;
  if (age > cfg.engine.sensor_offline_minutes) {
    return { action: "NO_DATA", minutes: null, reason: `sensor offline (${age.toFixed(0)} min since last reading)`, hours_until: null, trace };
  }
  const moisture = readings[0].moisture_pct;
  trace.push(`moisture=${moisture.toFixed(1)}%`);
  const wlogN = cfg.engine.waterlog_hours * 60 / 5;
  const wlogCount = readings.filter(r => r.moisture_pct >= cfg.engine.waterlog_pct).length;
  if (wlogCount >= wlogN) {
    return { action: "STOP", minutes: null, reason: "waterlogged, check drainage", hours_until: null, trace };
  }
  if (moisture >= mn) {
    const hrs = slopeHours(readings, mn, moisture) ?? et0Hours(weather, crop, moisture, mn);
    return { action: "OK", minutes: null, reason: `moisture ${Math.round(moisture)}% at or above minimum ${Math.round(mn)}%`, hours_until: hrs, trace };
  }
  const [mins, capped] = duration(moisture, tg, crop.water_holding_mm, zone.flow_mm_hr, eff, cfg.engine.max_run_minutes);
  const cap = capped ? " (split into cycles)" : "";
  const [rainEx, rainMm, rainProb] = rainExpected(weather, nowIso, cfg);
  const isCrit = moisture < crit;
  let override = "";
  if (rainEx && !isCrit) {
    return { action: "SKIP_RAIN", minutes: mins, reason: `rain forecast ${rainMm.toFixed(1)}mm ${Math.round(rainProb)}%${cap}`, hours_until: 0.0, trace, forecast_mm: rainMm, forecast_prob: rainProb };
  }
  if (rainEx && isCrit) override = " (critical, irrigating despite rain forecast)";
  if (zone.irrigation === "sprinkler" && weather && weather.wind_kmh > cfg.engine.wind_limit_kmh) {
    if (!isCrit) return { action: "WAIT_WIND", minutes: mins, reason: `wind ${weather.wind_kmh.toFixed(1)} km/h too high for sprinkler${cap}`, hours_until: null, trace };
    override = override || " (critical, irrigating despite wind)";
  }
  const [inWin, nextWin] = nextWindow(localMin, cfg.engine.windows);
  if (!inWin && !isCrit) {
    return { action: "IRRIGATE_LATER", minutes: mins, reason: `outside window, next at ${nextWin}${cap}`, hours_until: null, trace };
  }
  if (!inWin && isCrit) override = override || " (critical, irrigating outside window)";
  const reason = `moisture ${Math.round(moisture)}% is ${(mn - moisture).toFixed(1)} points below minimum ${Math.round(mn)}%${override}${cap}`;
  return { action: "IRRIGATE_NOW", minutes: mins, reason, hours_until: 0.0, trace };
}

File: sinchai/engine.py
from sinchai import clock


def _thresholds(zone, crop):
    mn = zone["min_pct"] if zone["min_pct"] is not None else crop["min_pct"]
    tg = zone["target_pct"] if zone["target_pct"] is not None else crop["target_pct"]
    return mn, tg


def _rain_expected(weather, now_iso, cfg):
    if weather is None:
        return False, 0.0, 0.0
    total, max_prob, count = 0.0, 0.0, 0
    for i, t in enumerate(weather["hourly_times"]):
        if t < now_iso:
            continue
        if count >= cfg["engine"]["rain_lookahead_hours"]:
            break
        total += weather["hourly_precip_mm"][i] or 0.0
        max_prob = max(max_prob, weather["hourly_precip_prob"][i] or 0.0)
        count += 1
    by_vol = total >= cfg["engine"]["rain_skip_mm"]
    by_prob = max_prob >= cfg["engine"]["rain_skip_prob"] and total >= cfg["engine"]["rain_skip_prob_min_mm"]
    return (by_vol or by_prob), total, max_prob


def _slope_hours(readings, mn, moisture):
    if len(readings) < 4:
        return None
    pts = [r["moisture_pct"] for r in reversed(readings[-12:])]
    n = len(pts)
    sx, sy = sum(range(n)), sum(pts)
    sxy = sum(i * pts[i] for i in range(n))
    sxx = sum(i * i for i in range(n))
    denom = n * sxx - sx * sx
    if denom == 0:
        return None
    slope = (n * sxy - sx * sy) / denom
    if slope >= 0:
        return None
    return (moisture - mn) / abs(slope) * 5.0 / 60.0


def _et0_hours(weather, crop, moisture, mn):
    if weather is None:
        return None
    vals = [v for v in weather["hourly_evaporation_mm"] if v is not None]
    if not vals:
        return None
    rate = sum(vals) / len(vals) * crop["crop_factor"] / crop["water_holding_mm"] * 100.0
    return (moisture - mn) / rate if rate > 0 else None


def _duration(moisture, tg, wh_mm, flow, eff, max_min):
    mins = (tg - moisture) / 100.0 * wh_mm / (flow * eff) * 60.0
    if mins <= 0:
        return 0.0, False
    return (max_min, True) if mins > max_min else (mins, False)


def _next_window(local_min, windows):
    parsed = [(int(w[:2]) * 60 + int(w[3:5]), int(w[6:8]) * 60 + int(w[9:11])) for w in windows]
    for s, e in parsed:
        if s <= local_min <= e:
            return True, None
    nxt = next((s for s, e in sorted(parsed) if s > local_min), min(p[0] for p in parsed))
    return False, f"{nxt // 60:02d}:{nxt % 60:02d}"


def decide(zone, crop, readings, weather, now_iso, local_min, cfg):
    if crop is None:
        raise ValueError(f"Zone {zone['id']}: unknown crop")
    mn, tg = _thresholds(zone, crop)
    if mn >= tg:
        raise ValueError(f"Zone {zone['id']}: min_pct must be less than target_pct")
    crit = mn - cfg["engine"]["critical_margin_pts"]
    trace = [f"min={mn}% target={tg}% critical={crit:.1f}%"]
    eff = cfg["efficiency"][zone["irrigation"]]

    if not readings:
        return {"action": "NO_DATA", "minutes": None, "reason": "no sensor data", "hours_until": None, "trace": trace}
    age = (clock.from_iso(now_iso) - clock.from_iso(readings[0]["ts"])).total_seconds() / 60.0
    if age > cfg["engine"]["sensor_offline_minutes"]:
        return {"action": "NO_DATA", "minutes": None, "reason": f"sensor offline ({age:.0f} min since last reading)", "hours_until": None, "trace": trace}

    moisture = readings[0]["moisture_pct"]
    trace.append(f"moisture={moisture:.1f}%")
    wlog_n = cfg["engine"]["waterlog_hours"] * 60 / 5
    if sum(1 for r in readings if r["moisture_pct"] >= cfg["engine"]["waterlog_pct"]) >= wlog_n:
        return {"action": "STOP", "minutes": None, "reason": "waterlogged, check drainage", "hours_until": None, "trace": trace}

    if moisture >= mn:
        hrs = _slope_hours(readings, mn, moisture) or _et0_hours(weather, crop, moisture, mn)
        return {"action": "OK", "minutes": None, "reason": f"moisture {moisture:.0f}% at or above minimum {mn:.0f}%", "hours_until": hrs, "trace": trace}

    mins, capped = _duration(moisture, tg, crop["water_holding_mm"], zone["flow_mm_hr"], eff, cfg["engine"]["max_run_minutes"])
    cap = " (split into cycles)" if capped else ""
    rain_ex, rain_mm, rain_prob = _rain_expected(weather, now_iso, cfg)
    is_crit = moisture < crit
    override = ""

    if rain_ex and not is_crit:
        return {"action": "SKIP_RAIN", "minutes": mins, "reason": f"rain forecast {rain_mm:.1f}mm {rain_prob:.0f}%{cap}", "hours_until": 0.0, "trace": trace}
    if rain_ex and is_crit:
        override = " (critical, irrigating despite rain forecast)"

    if zone["irrigation"] == "sprinkler" and weather is not None and weather["wind_kmh"] > cfg["engine"]["wind_limit_kmh"]:
        if not is_crit:
            return {"action": "WAIT_WIND", "minutes": mins, "reason": f"wind {weather['wind_kmh']:.1f} km/h too high for sprinkler{cap}", "hours_until": None, "trace": trace}
        override = override or " (critical, irrigating despite wind)"

    in_win, next_win = _next_window(local_min, cfg["engine"]["windows"])
    if not in_win and not is_crit:
        return {"action": "IRRIGATE_LATER", "minutes": mins, "reason": f"outside window, next at {next_win}{cap}", "hours_until": None, "trace": trace}
    if not in_win and is_crit:
        override = override or " (critical, irrigating outside window)"

    reason = f"moisture {moisture:.0f}% is {mn - moisture:.1f} points below minimum {mn:.0f}%{override}{cap}"
    return {"action": "IRRIGATE_NOW", "minutes": mins, "reason": reason, "hours_until": 0.0, "trace": trace}
